src/babashka/pods/impl.clj

```diff
@@ -310,59 +307,93 @@
                    (let [s (slurp f)]
                      (when (str/ends-with? s "\n")
                        (str/trim s))))]
-        (Integer. s)
+        (Integer/parseInt s)
         (recur)))))
 
 (defn debug [& strs]
   (binding [*out* *err*]
     (println (str/join " " (map pr-str strs)))))
 
+(defn resolve-pod [pod-spec {:keys [:version :force] :as opts}]
+  (let [resolved (when (qualified-symbol? pod-spec)
+                   (resolver/resolve pod-spec version force))
+        opts (if resolved
+               (if-let [extra-opts (:options resolved)]
+                 (merge opts extra-opts)
+                 opts)
+               opts)
+        pod-spec (cond
+                   resolved [(:executable resolved)]
+                   (string? pod-spec) [pod-spec]
+                   :else pod-spec)]
+    {:pod-spec pod-spec, :opts opts}))
+
+(defn run-pod [pod-spec {:keys [:transport] :as _opts}]
+  (let [pb (ProcessBuilder. ^java.util.List pod-spec)
+        socket? (identical? :socket transport)
+        _ (if socket?
+            (.inheritIO pb)
+            (.redirectError pb java.lang.ProcessBuilder$Redirect/INHERIT))
+        _ (cond-> (doto (.environment pb)
+                    (.put "BABASHKA_POD" "true"))
+                 socket? (.put "BABASHKA_POD_TRANSPORT" "socket"))
+        p (.start pb)
+        port-file (when socket? (port-file (.pid p)))
+        socket-port (when socket? (read-port port-file))
+        [socket stdin stdout]
+        (if socket?
+          (let [^Socket socket
+                (loop []
+                  (if-let [sock (try (create-socket "localhost" socket-port)
+                                     (catch java.net.ConnectException _
+                                       nil))]
+                    sock
+                    (recur)))]
+            [socket
+             (.getOutputStream socket)
+             (PushbackInputStream. (.getInputStream socket))])
+          [nil (.getOutputStream p) (java.io.PushbackInputStream. (.getInputStream p))])]
+    {:process p
+     :socket socket
+     :stdin stdin
+     :stdout stdout}))
+
+(defn describe-pod [{:keys [:stdin :stdout]}]
+  (write stdin {"op" "describe"
+                "id" (next-id)})
+  (read stdout))
+
+(defn describe->ops [describe-reply]
+  (some->> (get describe-reply "ops") keys (map keyword) set))
+
+(defn describe->metadata [describe-reply]
+  (let [format (-> (get describe-reply "format") bytes->string keyword)
+        ops (describe->ops describe-reply)
+        readers (when (identical? :edn format)
+                  (read-readers describe-reply resolve))]
+    {:format format, :ops ops, :readers readers}))
+
+(defn load-pod-metadata [pod-spec opts]
+  (let [{:keys [:pod-spec :opts]} (resolve-pod pod-spec opts)
+        running-pod (run-pod pod-spec opts)
+        describe-reply (describe-pod running-pod)
+        ops (describe->ops describe-reply)]
+    (future (destroy* (assoc running-pod :ops ops)))
```

Comment:
I don't think we should do this in a future, as we might kill the pod before it has been shut down.

Reply:
How so? All killing / shutting down happens inside the future. I added this so we don't have to wait for the shutdown to complete before returning the describe reply.

Reply:
If babashka gets killed before the future ends, the pod may be killed prematurely.

Reply:
I thought it waited on all futures to complete unless you called `shutdown-agents`. Maybe that's different in babashka?

Would it help if I added a shutdown hook to these pod processes too (like `load-pod` does)?

Reply:
babashka does call `shutdown-agents` in main. Let's not mess with async code if we don't have to. What is the problem that is solved by using a future? This is the uncached case and should be relatively rare. Besides, shutting down the pod should only take a few milliseconds or less.

Reply:
Gotcha. The problem was that a pod-supplied shutdown op could do anything and take any amount of time it wants to do it. And it provides nothing we need to move forward with what we're trying to accomplish.

But I'll take it out if you don't think it's worth backgrounding.

Reply:
Let's revisit if we feel it's necessary, I don't see strong evidence right now.